Sources/GRPCCore/Metadata.swift

```diff
@@ -14,5 +14,267 @@
+  ///   - key: The key to be associated with the given value. Must end in `-bin`.
+  ///   - binaryValue: The `Data` to be associated with the given key.
+  public mutating func add(key: String, binaryValue: Data) {
+    self.add(key: key, value: .binary(binaryValue))
+  }
+  
+  /// Add a new key-value pair.
+  /// - Parameters:
+  ///   - key: The key to be associated with the given value. If value is binary, it must end in `-bin`.
+  ///   - value: The ``MetadataValue`` to be associated with the given key.
+  public mutating func add(key: String, value: MetadataValue) {
+    self.elements.append(.init(key: key, value: value))
+  }
+  
+  /// Adds a key-value pair to the collection, where the value is a string.
+  /// If there are pairs already associated to the given key, they will all be removed first, and the new pair
+  /// will be added. If no pairs are present with the given key, a new one will be added.
+  /// - Parameters:
+  ///   - key: The key to be associated with the given value.
+  ///   - stringValue: The string value to be associated with the given key.
+  public mutating func replaceOrAdd(key: String, stringValue: String) {
+    self.lockedElements.withLockedValue { elements in
+      elements.removeAll { metadataKeyValue in
+        metadataKeyValue.key == key
+      }
+      elements.append(.init(key: key, value: .string(stringValue)))
+    }
+  }
+  
+  /// Adds a key-value pair to the collection, where the value is `Data`.
+  /// If there are pairs already associated to the given key, they will all be removed first, and the new pair
+  /// will be added. If no pairs are present with the given key, a new one will be added.
+  /// - Parameters:
+  ///   - key: The key to be associated with the given value. Must end in `-bin`.
+  ///   - binaryValue: The `Data` to be associated with the given key.
+  public mutating func replaceOrAdd(key: String, binaryValue: Data) {
+    self.lockedElements.withLockedValue { elements in
+      elements.removeAll { metadataKeyValue in
+        metadataKeyValue.key == key
+      }
+      elements.append(.init(key: key, value: .binary(binaryValue)))
+    }
+  }
+  
+  /// Adds a key-value pair to the collection.
+  /// If there are pairs already associated to the given key, they will all be removed first, and the new pair
+  /// will be added. If no pairs are present with the given key, a new one will be added.
+  /// - Parameters:
+  ///   - key: The key to be associated with the given value. If value is binary, it must end in `-bin`.
+  ///   - value: The ``MetadataValue`` to be associated with the given key.
+  public mutating func replaceOrAdd(key: String, value: MetadataValue) {
+    self.lockedElements.withLockedValue { elements in
+      elements.removeAll { metadataKeyValue in
+        metadataKeyValue.key == key
+      }
+      elements.append(.init(key: key, value: value))
+    }
+  }
+}
+
+extension Metadata: RandomAccessCollection {
+  public typealias Element = MetadataKeyValue
```

Comment:
I think we should just use a named tuple here, it's a bit more ergonomic as you can destructure it in a loop (e.g. `for (key, value) in metadata { ... }`) and what users would expect from iterating key-value types like dictionary.

Reply:
I originally tried this, but when I define a typealias
```swift
public typealias KeyValue = (key: String, value: Value)
```
I get an error that `Metadata` doesn't conform to `Equatable` for some reason (`Value` is `Hashable`).
I tried defining `==` manually and equating the `elements` array I got an error saying I couldn't compare two arrays of type `KeyValue aka (String, Value)`.
Is there some conformance or other obvious thing I'm missing?

Reply:
Oh yeah, you're totally right. Tuples don't get useful conformance out of the box. I think we can go halfway by storing the `struct` you defined above but still using the tuple as the `Element` type and just unpack the `struct` wherever we need to.